modules/tftp/server.rb

```diff
@@ -90,21 +107,93 @@ def pxe_default
     end
 
     def pxeconfig_file(mac)
-      ["#{pxeconfig_dir}/menu.lst.01" + mac.delete(':').upcase, "#{pxeconfig_dir}/01-" + mac.tr(':', '-').upcase]
+      ["#{pxeconfig_dir}/menu.lst.01" + mac.delete(':').upcase, "#{pxeconfig_dir}/01-" + dashed_mac(mac).upcase]
     end
   end
 
   class Pxegrub2 < Server
-    def pxeconfig_dir
-      "#{path}/grub2"
+    def bootloader_path(os, version, arch)
+      unless (bootloader_universe = Proxy::TFTP::Plugin.settings.bootloader_universe)
+        logger.debug "TFTP: bootloader universe not configured."
+        return
+      end
+
+      bootloader_path = File.join(bootloader_universe, 'pxegrub2', os, version, arch)
+
+      logger.debug "TFTP: Checking bootloader universe for suitable bootloader directory for"
+      logger.debug "TFTP:   * Operating system: #{os}"
+      logger.debug "TFTP:   * Version: #{version}"
+      logger.debug "TFTP:   * Architecture: #{arch}"
+      logger.debug "TFTP: Checking bootloader universe if \"#{bootloader_path}\" exists."
+      unless Dir.exist?(bootloader_path)
+        logger.debug "TFTP: Directory \"#{bootloader_path}\" does not exist."
+
+        bootloader_path = "#{bootloader_universe}/pxegrub2/#{os}/default/#{arch}"
+        logger.debug "TFTP: Checking if fallback directory at \"#{bootloader_path}\" exists."
+        unless Dir.exist?(bootloader_path)
+          logger.debug "TFTP: Directory \"#{bootloader_path}\" does not exist."
+          return
+        end
+      end
+
+      bootloader_path
+    end
+
+    def setup_bootloader(mac, os, major, minor, arch, bootfilename_efi, build)
```

Comment:
This is my `universe` directory
```
:universe/pxegrub2/fedora/39/x86_64/
├── grubx64.efi
└── shimx64.efi
```

When I create a host, the `host_config` dir structure is prepared like this:
```
host_config/
└── 0a-1a-2a-3a-4a-5a
    └── grub2
        ├── grub.cfg
        ├── grub.cfg-01-0a-1a-2a-3a-4a-5a
        ├── grub.cfg-0a:1a:2a:3a:4a:5a
        ├── grubx64.efi
        ├── os_info
        └── shimx64.efi
```

The DHCP lease looks like this:
```
host wayne-bartucca.dhcp.lan {
  dynamic;
  hardware ethernet 0a:1a:2a:3a:4a:5a;
  fixed-address 192.168.180.248;
        supersede server.filename = 
                "host_config/0a-1a-2a-3a-4a-5a/grub2/boot.efi";
        supersede server.next-server = c0:a8:be:01;
        supersede host-name = "wayne-bartucca.dhcp.lan";
```

As you can see, the filename points to the `host_config/0a-1a-2a-3a-4a-5a/grub2/boot.efi` file,
but the `boot.efi` file is not there. Because of that, the provisioning will fail.

Reply:
Hey Leos :wave: You have to manually create the symlinks of `boot.efi` to `grubx64.efi` and `boot-sb.efi` to `shimx64.efi` in the `universe/pxegrub2/fedora/39/x86_64/` directory (see [step 6 of the Procedure for Rocky Linux](https://theforeman-foreman-documentation-preview-pr-2145.surge.sh/nightly/Provisioning_Hosts/index-katello.html#configuring-smart-proxy-to-provision-rocky-on-Secure-Boot-enabled-hosts) for comparison).

Reply:
JFYI: the provision of these files could then also be taken over [nboci-files](https://github.com/theforeman/nboci-files) in the future.